fix(components): [date-picker] Fix the problem that clicking the month button automatically selects unavailable dates

1120040377 opened the pull request:
closed #10604

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

Commit:
fix(components): issues/10604 如当月设置禁用1号后，仍然可以通过顶部月份入口默认选中当月1号

closed #10604

Commit:
fix(components): issues/10604 如当月设置禁用1号后，仍然可以通过顶部月份入口默认选中当月1号

closed #10604

chenxch commented on packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:
```diff
@@ -396,7 +396,9 @@ const handleMonthPick = async (month: number) => {
-      emit(innerDate.value, true)
+      if (!disabledDate || !disabledDate(innerDate?.value?.toDate())) {
+        emit(innerDate.value, true)
+      }
```
Maybe we should cancel this selected operation.

1120040377 replied:
At first, I thought so too.

It was later discovered that not all dates of the current month cannot be selected. If users disable dates before November 15th, they will still need to click the November tab to select a different date for that month.

Since the disabled date is determined by the function, if you want to automatically select the first optional date, you need to traverse all the days of the month to achieve.

So in the end I opted to skip the automatic allocation

1120040377 commented:
> Hello @1120040377 , Please modify the reasonable title.

[Element Plug Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCBtc2cgPSByZWYoJ0hlbGxvIFdvcmxkIScpXG5jb25zdCB2YWx1ZTEgPSByZWYoJycpXG5jb25zdCBzZGQgPSAodGltZTogRGF0ZSkgPT4ge1xuICByZXR1cm4gdGltZS5nZXRUaW1lKCkgPCBEYXRlLm5vdygpXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8aDE+e3sgbXNnIH19PC9oMT5cbiAgPGVsLWlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuXG4gIDxwPlxuICAgIDxlbC1kYXRlLXBpY2tlclxuICAgICAgICAgIHYtbW9kZWw9XCJ2YWx1ZTFcIlxuICAgICAgICAgIHR5cGU9XCJkYXRlXCJcbiAgICAgICAgICBmb3JtYXQ9XCJNTS9ERC9ZWVlZXCJcbiAgICAgICAgICA6ZGlzYWJsZWQtZGF0ZT1cInNkZFwiXG4gICAgICAgICAgcGxhY2Vob2xkZXI9XCJQaWNrIGEgZGF5XCJcbiAgICAgICAgLz5cbiAgICBFbGVtZW50IFBsdXMge3sgRXBWZXJzaW9uIH19ICsgVnVlIHt7IHZ1ZVZlcnNpb24gfX1cbiAgPC9wPlxuPC90ZW1wbGF0ZT5cblxuIiwiUGxheWdyb3VuZE1haW4udnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9wcmV2aWV3LTEwNjE0LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIlxuICB9XG59IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtZG9tQGxhdGVzdC9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdnVlL3NoYXJlZFwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTA2MTQtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgIGxpbmsuaHJlZiA9ICdodHRwczovL3ByZXZpZXctMTA2MTQtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MnXG4gICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kKGxpbmspXG4gIH0pXG59IiwiX28iOnsic2hvd0hpZGRlbiI6dHJ1ZSwic3R5bGVTb3VyY2UiOiJodHRwczovL3ByZXZpZXctMTA2MTQtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MifX0=)

Commit:
fix(components): [date-picker] Fix automatically selects date

Fix the problem that clicking the month button automatically selects unavailable dates

closed #10614

chenxch commented:
I want to say that the action of switching months should either not select the optional first day at all, or select both, and the interaction should be unified. (I personally prefer to say that it is not selected, because element-ui does the same

1120040377 commented:
After seeing your feedback, I went to the documentation of Element, and it is true that the date is not automatically selected.

I don't know why the datepicker in Element+ was originally designed like this, it may be for user experience, because in some cases, users may only want to change the year or month, and it may be more convenient to click the tab operation. (Maybe we can communicate, how to modify the experience here will be better)

1120040377 commented:
Same problem with DateTimePicker

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
fix(components): [date-picker] bug fix

Fix the problem that clicking the month button automatically selects unavailable dates

closed #10614

1120040377 committed:
Merge branch 'dev' into dev

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge branch 'dev' of github.com:1120040377/element-plus into dev

Baymi commented:
Is this problem solved？Why this merge is blocked

1120040377 commented:
I don't know. Nobody answered me



---Original---
From: ***@***.***&gt;
Date: Thu, Dec 1, 2022 14:35 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [element-plus/element-plus] fix(components): [date-picker] Fixthe problem that clicking the month button automatically selects unavailabledates (PR #10614)




 
Is this problem solved？Why this merge is blocked
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

tolking commented:
Thanks for your contribution.

This issues has been fixed.
